### ISSUE TITLE
Bbox Pruning Too Aggressive?

### ISSUE THREAD

**Issue by 5had3z** [NVIDIA/DALI#5366]
### Version

1.34.0

### Describe the bug.

More bounding boxes are removed than expected around the edges of an image with a random crop. A few examples of the full aug pipeline and one with no augs is shown below. I don't think flipping and scaling have much to do with this. I tried to play with thresholds to allow for boxes that have some part in the ROI to be retained, but I don't think really loose `thresholds` and `threshold_type` seems to help. I don't think they relate to culling boxes, rather that they ensure some box is present, but I tried anyway.

- Example 1 
![aachen_000040_000019](https://github.com/NVIDIA/DALI/assets/59598353/b18775b9-373a-45b2-b8f2-7596be31cebf)
![aachen_000040_000019_full](https://github.com/NVIDIA/DALI/assets/59598353/d6f8c4b3-e887-410e-b719-e4de3bfe5cba)

- Example 2
![aachen_000112_000019](https://github.com/NVIDIA/DALI/assets/59598353/81145c40-7ac3-42c9-bb2f-5bf5a9043caa)
![aachen_000112_000019_full](https://github.com/NVIDIA/DALI/assets/59598353/7712a6f3-b4da-438e-9351-1b6e9f96764a)

### Minimum reproducible example

```shell
# Excerpt from data pipeline with cropping and flipping, since bbox is relative
# coords, scale has no effect and no change to bbox is made. Padding data and display
# results snippets are also shown.

### Cropping Snippet
keys = list(filter(_is_image_like, pipedata))
img_shape = fn.shapes(pipedata[keys[0]], dtype=DALIDataType.INT32)[-2::-1]
anchor, shape, pipedata["bboxes"], pipedata["labels"], bboxes_indicies = (
    fn.random_bbox_crop(
        pipedata["bboxes"],
        pipedata["labels"],
        input_shape=img_shape,
        crop_shape=size,
        bbox_layout="xyXY",
        output_bbox_indices=True,
        thresholds=0.0,
        threshold_type="overlap",
    )
)
outputs = fn.slice(
    [pipedata[k] for k in keys],
    anchor,
    shape,
    normalized_anchor=False,
    normalized_shape=False,
)

### Flipping snippet
targets = list(filter(_is_image_like, pipedata))
target_data = [pipedata[k] for k in targets]
should_flip = fn.random.coin_flip(probability=0.5, dtype=DALIDataType.INT32)
target_data = fn.flip([pipedata[k] for k in targets], horizontal=should_flip)
inplace_update(pipedata, targets, target_data)
pipedata["bboxes"] = fn.bb_flip(
    pipedata["bboxes"], horizontal=should_flip, ltrb=True
)

# Pad uneven samples
pipedata["bboxes"] = fn.pad(pipedata["bboxes"], fill_value=-1)
pipedata["labels"] = fn.pad(pipedata["labels"], fill_value=-1)

# Show results snippet
def draw_bboxes_dali(data: dict[str, Tensor]):
    bboxes = data["bboxes"][0][data["labels"][0] != -1]
    h, w = data["image"].shape[2:]
    bboxes *= torch.tensor([w, h, w, h]).cuda()
    image = (data["image"][0] * 255).to(torch.uint8)
    names = [trainId2label[l.item()].name for l in data["labels"][0] if l.item() != -1]
    drawn = draw_bounding_boxes(image, bboxes, colors=(255, 0, 0), labels=names)
    fname = filename_tensor_2_str(data["filename"])[0]
    plt.title(fname)
    plt.imshow(drawn.permute(1, 2, 0).cpu().numpy())
    plt.savefig(fname + "_full")
    plt.close("all")
```

### Other/Misc.

As a side, maybe adding keys=HW to fn.shapes for named dimensions for convenience would be good (instead of the fn.shapes mess in the code).
Querying "ndim" is also a missing feature (maybe I have FCHW clip or CHW image and want to do something different for either).

### Check for duplicates

- [X] I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by jantonguirao** [comment]
@5had3z Thank you for reporting this issue. To make sure I understand the problem, could you also provide some toy example data and code snippet so that I can reproduce and debug on my end? You can also play with pipeline seed to provide reproducible results. 
Thank you in advance.

**Comment by jantonguirao** [comment]
> More bounding boxes are removed than expected around the edges of an image with a random crop. 

Can you also explain, why you mean by this? What do we mean by "expected". This operator removes any bounding box with a centroid outside of the cropping region. That is, if the center of the box is outside of the random crop, it's filtered out. Is this what you mean?

**Comment by 5had3z** [comment]
> 
> From the documentation:
> 
> ```
> The third output contains the bounding boxes, after filtering out the ones with a centroid outside
> of the cropping window, and with the coordinates mapped to the new coordinate space.
> ```

I should've rtfm'd a bit harder, I mainly focused on using the anchor, shape and output_bbox_indices output correctly. Since this is currently expected behaviour, I think a new parameter should be added to prune boxes based on area of bbox within the ROI, which I would argue would a prefered metric.

With the current behaviour if my centroid is in the ROI by one pixel in both axis, I have ~25% of the original box and this is kept. But if my bbox centroid is outside of the ROI by 1 pixel in 1 axis, I have just lost a bbox that is still ~50% within the image. Using a %area still within the ROI would keep more boxes that  are reasonably within the ROI and would enable the option to make it 0.0 for no pruning.

I'll copy-pasta the existing code and add this new behaviour to a custom plugin for myself for now. If there is interest I can then make the changes on the main fn.random_bbox_crop.

**Comment by 5had3z** [comment]
Comparison with my new cropping algorithm. This is shown with threshold=0.0 so if any sliver of the box is within roi it is kept, but this can be tuned. Not much needed to be changed, new arg and `FilterByCentroid` replaced with `FilterBboxes` that takes a  lambda that returns a "should_keep" bool based on crop and box.

-Example 1
**Centroid**
![aachen_000000_000019_new_centroid](https://github.com/NVIDIA/DALI/assets/59598353/91cb506a-f64f-432c-bb90-b37e27dff4e7)
**Threshold**
![aachen_000000_000019_new_thresh](https://github.com/NVIDIA/DALI/assets/59598353/6ce5f4e3-71f7-465a-974d-9d09b2df94e6)

-Example 2
**Centroid**
![aachen_000040_000019_new_centroid](https://github.com/NVIDIA/DALI/assets/59598353/43aa3b38-b125-4308-b65a-4eca30dbe9a8)
**Threshold**
![aachen_000040_000019_new_thresh](https://github.com/NVIDIA/DALI/assets/59598353/f675dc06-e6db-427c-a7ad-795790e250f0)

-Example 3
**Centroid**
![aachen_000080_000019_new_centroid](https://github.com/NVIDIA/DALI/assets/59598353/afe83a92-b7df-4119-9a5c-4a14923413ce)
**Threshold**
![aachen_000080_000019_new_thresh](https://github.com/NVIDIA/DALI/assets/59598353/a84d3d95-d8fc-4ccc-9fa1-1f53095f25c8)
